Define intermediate format (#1058)
Proteobench/ProteoBench: docs/developer-guide/intermediate-format.md

# Intermediate format specification

This document specifies the internal tabular formats that ProteoBench produces
while processing a benchmark submission. These formats are for the internal use of:
scoring, plotting, datapoint generation, the submission-validation layer, and the
`intermediate_hash` that identifies a dataset all depend on the column names,

> types, and semantics defined here.

Reviewer: maybe we can add here that these are module specific?

```{note}
Status: descriptive specification of the current behaviour (format version 1,
implicit). It documents what the code produces today. Proposed changes (an
explicit version field, a canonical serialisation for hashing) are described in
the "Reproducibility and the intermediate hash" and "Versioning" sections below.
```

## Scope

There are two distinct tables:

| Artifact | Produced by | Persisted | Consumed by |
|----------|-------------|-----------|-------------|
| **Standard format** (pre-scoring) | `ParseSettingsQuant.convert_to_standard_format()` | No (in-memory) | Scoring; the submission-validation layer |
| **Intermediate format** (post-scoring) | `QuantScoresHYE.generate_intermediate()` | Yes, as `result_performance.csv` in the dataset archive; also hashed into `intermediate_hash` | Plotting, datapoint metrics, the "View Single Result" table |

The intermediate format is the primary subject of this specification. The
standard format is documented as the declared input format that any module or
tool parser must produce.

## Standard format

Produced by `convert_to_standard_format()` in
`proteobench/io/parsing/parse_settings.py`. It is a long table with one row per

> (precursor, run): a precursor that was quantified in six runs yields six rows.

Reviewer: here, "precursor, run" is module specific, right? 
I suspect that the specification of this standard format should be module specific, no?

Reviewer: I propose for lines 31-36: 

It is a long table with one row per measured feature (for example precursor, run for the precursor quantification modules). Here, we describe the columns that are present in the current modules, they are not all needed in all modules, and unmapped input columns may also be present.

And I also propose that we add a column in the table that states what module uses what field?

The table is not projected to a fixed column set, so unmapped input columns may
also be present; only the columns below are relied upon downstream (this is quite specific
for quantification modules; may differ for other modules!).

| Column | Type | Required | Meaning |
|--------|------|----------|---------|
| `Proteins` | str | yes | Protein or protein-group identifier(s). Multiple proteins may be joined by `;` or `,`; UniProt-style identifiers are split on the pipe character into database, accession, and entry name. |
| `Sequence` | str | yes | Plain (unmodified) peptide sequence. |
| `Charge` | int | yes | Precursor charge. |

> | `proforma` | str | conditional | ProForma-encoded modified sequence. Present only when the tool TOML defines a `[modifications_parser]`. |

Reviewer: ```suggestion
| `proforma` | str | conditional | [ProForma](https://github.com/HUPO-PSI/ProForma)-encoded modified sequence. Present only when the tool TOML defines a `[modifications_parser]`. |
```

| `precursor ion` | str | conditional | `proforma + "/" + Charge`. Present when `level = "ion"`. |
| `peptidoform` | str | conditional | Equal to `proforma`. Present when `level = "peptidoform"`. |
| `Raw file` | str | yes | Cleaned run name (extensions and known suffixes stripped by `_clean_run_name()`). |
| `Intensity` | float | yes | Quantitative intensity for the (precursor, run). Rows with `Intensity <= 0` are removed. |
| `replicate` | str | yes | Condition label, `"A"` or `"B"`, mapped from `Raw file` via the tool's `condition_mapper`. |
| `contaminant` | bool | yes | Result of matching `Proteins` against the contaminant flag. Contaminant rows are removed, so survivors are all `False`. |
| species flags | bool | yes | One boolean column per species in the module `species_mapper` (for HYE: `HUMAN`, `YEAST`, `ECOLI`), set by substring match on `Proteins`. |
| `MULTI_SPEC` | bool | yes | `True` when the precursor matches more than `min_count_multispec` species. Multi-species rows are removed. |
| run dummy columns | bool | yes | One boolean column per distinct `Raw file` (from `pd.get_dummies`). |

## Intermediate format

> The intermediate format is the table defined for calculating all metrics. It should contain

Reviewer: ```suggestion
The intermediate format is the table defined for calculating all metrics in a given module. It should contain
```

every value needed for metric calculation in a standardized fashion. It is not limited to 
these columns, it can contain more columns, even ones that are tool specific. However, it at

> least needs defined and "normalized" columns for metric calculation.

Reviewer: I don't understand what "defined and "normalized"" mean here.

> An example for quantificaiton modules; produced by `QuantScoresHYE.generate_intermediate()` in

Reviewer: proposition for the lines 61-65:

Depending on the module and the software tool, some filtering steps can be applied to the data before being stored in the intermediate format. 

We present bellow the intermediate format tables produced by QuantScoresHYE.generate_intermediate() in proteobench/score/quantscoresHYE.py (mixed-species quantification modules). These contain one row per precursor (or peptidoform), and only single-species precursors are retained (compute_epsilon() keeps rows where exactly one species flag is set (unique == 1), so multi-species precursors do not appear). The intermediate formats for other modules are described in the section Module Variants soon after.

`proteobench/score/quantscoresHYE.py`. One row per precursor (or peptidoform).
Only **single-species** precursors are retained: `compute_epsilon()` keeps rows
where exactly one species flag is set (`unique == 1`), so multi-species
precursors do not appear.

### Column catalog

| Column | Type | Meaning |
|--------|------|---------|

> | `precursor ion` or `peptidoform` | str | Precursor key. Name depends on the module `level`. For ion modules this is `proforma + "/" + charge`. |

Reviewer: shouldn't we use "Feature key" instead of "Precursor key" since it is now precursor or peptidoform and may be more different things in the future?

| `Intensity_mean_A`, `Intensity_mean_B` | float | Mean linear intensity across the runs of condition A / B. |
| `Intensity_std_A`, `Intensity_std_B` | float | Standard deviation of linear intensity per condition. |
| `log_Intensity_mean_A`, `log_Intensity_mean_B` | float | Mean of the log2 intensities per condition. |
| `log_Intensity_std_A`, `log_Intensity_std_B` | float | Standard deviation of log2 intensities per condition. |
| `CV_A`, `CV_B` | float | Coefficient of variation per condition, `Intensity_std / Intensity_mean`. |
| `log2_A_vs_B` | float | `log_Intensity_mean_A - log_Intensity_mean_B`. The observed log2 fold change. |

> | per-run intensity columns | float | One column per distinct run name (cleaned `Raw file`), holding that run's intensity for the precursor. The set of columns therefore depends on the experimental design. |

Reviewer: ```suggestion
| per-run intensity columns | float | One column per distinct run name (cleaned `Raw file`), holding that run's intensity for the quantified feature. The set of columns therefore depends on the experimental design. |
```

> | `nr_observed` | int | Number of runs in which the precursor was quantified. |

Reviewer: ```suggestion
| `nr_observed` | int | Number of runs in which the feature (i.e. precursor, peptidoform) was quantified. |
```

> | species flags | bool | One boolean per species (HYE: `YEAST`, `ECOLI`, `HUMAN`). For a retained row exactly one is `True`. |

Reviewer: ```suggestion
| species flags | bool | One boolean per species (HYE: `YEAST`, `ECOLI`, `HUMAN`). Rows with more than one `True` values are excluded from the table. |
```

- the **row order**;
- the floating-point formatting used by `to_string()`;
- the run-specific intensity column names.

Any change to the scoring code, the column order, or the pandas rendering can
change the hash for identical inputs. Implementers must therefore treat the
column order and row order as part of the format.

## Module variants

- **HYE (Human-Yeast-Ecoli)**: the format described above. Species set
  `HUMAN`, `YEAST`, `ECOLI`.
- **PYE (Plasma, `quant_lfq_DIA_ion_plasma`)**: `QuantScoresPYE.generate_intermediate()`
  is currently a pass-through over the HYE implementation, so the intermediate
  columns are the same. The plasma-specific quantities (spike-in fold-change
  error, dynamic range, human-plasma epsilon) are computed later, in
  `QuantDatapointPYE`, not in the intermediate table. The plasma datapoint also
  uses `max_nr_observed = 12` instead of 6.
- **De novo (`denovo_DDA_HCD`)**: a different schema entirely, produced by
  `DenovoScores.generate_intermediate()`. It is keyed on spectrum and peptidoform
  and carries match-type and amino-acid-level columns (for example `match_type`,
  `aa_matches_dn`/`aa_matches_gt`, `pep_match`) rather than the quantitative

>   columns above. A detailed catalogue for the de novo intermediate is out of

Reviewer: ```suggestion
  columns above. This modules is still in development, the full documentation of its intermediate format will be provided soon.
```

Reviewer: ideally before it switches to BETA

>   scope for this revision and should be added alongside the de novo module's

Reviewer: ```suggestion
```

>   maturation.

Reviewer: ```suggestion
```

## Persistence and external resources

> The intermediate is written to `result_performance.csv` inside the dataset

Reviewer: ```suggestion
The intermediate format table is written to `result_performance.csv` inside the dataset
```
